### ISSUE TITLE
Cannot multiply reference external context

### ISSUE THREAD

**Issue by breyed** [EntityGraphQL/EntityGraphQL#206]
If a query references a DB context that is distinct from the EntityGraphQL context more than once, the query fails with something like:

> Field 'users' - variable 'db' of type 'EntityGraphQL.Tests.TestDataContext2' referenced from scope '', but it is not defined

It is common to reference a context multiple times when joining additional tables. Implicit joins made by EF via navigation properties are OK. It's only when the context appears more than once in the query that the error occurs.

A unit test is in #201 commit 434d46ed4524f41d3228417eea0e135049fe05b3.

**Comment by lukemurray** [comment]
Note that this fixes the issue of the incorrect expression being built. It may not work in the way you expect... By default EnitytGraphQL executes an expression to select all the fields from the base/main query context. In the case the main query context is a DbContext that means it'll execute the expression, triggering EF and all that. Then use the result with the next service field expressions.

Basically I believe it will work but in this case will be 2 calls to the DB. I actually don't know if EF normally would make that 1 call or not? I guess the 2 DbContext's could be different DBs? It's not a pattern I'm familiar with.

**Comment by breyed** [comment]
Here's the use case that led to finding this issue:

I have an existing DbContext used by the web app. Only a customized subset should be exposed over GraphQL, especially at the top level. I created a class that doesn't derive from DbContext but instead takes a reference to the DbContext through DI. It contains the properties intended to be exposed through GraphQL and is the generic type for a call to `SchemaBuilder.FromObject`.

For one of the tables in the DbContext exposed through GraphQL a few levels into the graph, I added a field whose IQueryable return value has joins that use the DbContext. Since the IQueryable doesn't rely solely on navigation properties of the type the field is being added to, I call `ResolveWithService` and pass the DbContext as the service. When the DbContext was referenced multiple times, I encountered the issue.

In ordinary EF, you can reference the DbContext multiple times and still end up with a single query.
```c#
from order in db.Orders
where order.Date > startDate
join detail in db.Details on order equals detail.Order
select new { order.Title, detail.Name }
```
should be equivalent to
```c#
from detail in db.Details
where detail.Order.Date > startDate
select new { detail.Order.Title, detail.Name }
```

In either case, the number of queries should depend on what you pass to `UseQuerySplittingBehavior()` at configuration time, which tells EF whether to use 1 or 2 queries to solve the N+1 problem.

At least, that's how I understand it. But if you have any insights, please share. It would be good to flesh out any important concepts for the EntityGraphQL documentation.

**Comment by lukemurray** [comment]
https://entitygraphql.github.io/docs/entity-framework#how-entitygraphql-handles-resolvewithservice-with-ef this tries to explain how what I call "service fields" are resolved in EntityGraphQL.

This is a result of wanting EntityGraphQL not have a dependency on EF but having an options to split the execution of base query/schema context fields from service fields to support EF well. This option can be turned off (`ExecutionOptions.ExecuteServiceFieldsSeparately`) if you know it is not needed (EF often throws errors mixing EF fields and other fields that depending on the complexity).

Turning it off if you have no other service fields may then follow the EFs `UseQuerySplittingBehavior()`. I have not tested be great if you could report back (using master where the original error is fixed).

You also cloud look at using the 1 `DbContext` and either
1. Start with an empty schema and only add the fields and types you want. We do this for a documented API so we can have more control on changes to it (e.g. we can change the underlying DbContext fields., names or types but keep them exposed as they were until we can do a release documenting the breaking changes)
2. Use `GraphQLIgnoreAttribute`, `schema.RemoveType()`, `schema.Type<T>().RemoveField()` and friends to clean up things you do not want exposed. Issue with this one is it is easy to have new fields etc slip in

Or, assuming it is all available in one DbContext
- Expose the field you need for the join on that second DbContext but remove that one from the schema using the methods above. Then it is still available in the expressions you use to build other fields.

Hope that made sense!
